lambda/translcore.ml

```diff
@@ -242,13 +242,35 @@ and transl_exp0 ~in_new_scope ~scopes e =
       event_after ~scopes e
         (transl_apply ~scopes ~tailcall ~inlined ~specialised
            (transl_exp ~scopes funct) oargs (of_location ~scopes e.exp_loc))
-  | Texp_match(arg, pat_expr_list, partial) ->
+  | Texp_match(arg, pat_expr_list, [], partial) ->
       transl_match ~scopes e arg pat_expr_list partial
-  | Texp_try(body, pat_expr_list) ->
+  | Texp_match(arg, pat_expr_list, eff_pat_expr_list, partial) ->
+  (* need to separate the values from exceptions for tansl_handler *)
+      let split_case (val_cases, exn_cases as acc)
+            ({ c_lhs; c_rhs } as case) =
+        if c_rhs.exp_desc = Texp_unreachable then acc else
+        let val_pat, exn_pat = split_pattern c_lhs in
+        match val_pat, exn_pat with
+        | None, None -> assert false
+        | Some pv, None ->
+            { case with c_lhs = pv } :: val_cases, exn_cases
+        | None, Some pe ->
+            val_cases, { case with c_lhs = pe } :: exn_cases
+        | Some _pv, Some _pe ->
+            assert false (* FIXME: handle this case *)
```

Comment:
For the record: this case happens with cases such as `0 | exception Not_found -> e` .  Here is a trivial fix.  It duplicates the right-hand side `e`, but the only alternative I can see is to introduce a local function definition for `e`, which the user can do as well.  (In the non-effect-handling case, the compilation of cases like this uses static handlers to share `e`, which is more efficient than a local function definition, and something that users cannot write in the source code.)
```suggestion
        | Some pv, Some pe ->
            { case with c_lhs = pv } :: val_cases,
            { case with c_lhs = pe } :: exn_cases
```